bigchaindb/core.py

```diff
@@ -151,16 +150,74 @@ def get_transaction(self, txid):
             If no transaction with that `txid` was found it returns `None`
         """
 
-        response = r.table('bigchain').concat_map(lambda doc: doc['block']['transactions'])\
-            .filter(lambda transaction: transaction['id'] == txid).run(self.conn)
+        validity = self.get_blocks_status_containing_tx(txid)
+
+        if validity:
+            # Disregard invalid blocks, and return if there are no valid or undecided blocks
+            validity = {_id: status for _id, status in validity.items() if status != 'invalid'}
+            if not validity:
+                return None
+
+            # If the transaction is in a valid or any undecided block, return it. Does not check
+            # if transactions in undecided blocks are consistent, but selects the valid block before
+            # undecided ones
+            for _id in validity:
+                target_block_id = _id
+                if validity[_id] == 'valid':
+                    break
+
+            # Query the transaction in the target block and return
+            response = r.table('bigchain').filter(lambda block: block['id'] == target_block_id).nth(0)\
```

Comment:
can we use just `get`? The `id` is the primary index, so it's fast and the code is more compact. Also, we can get rid of the `nth(0)` operator.